File: quickload-core/src/main/java/org/quickload/record/Schema.java
package org.quickload.record;

import java.util.List;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

public class Schema
{
    private final List<Column> columns;

    @JsonCreator
    public Schema(List<Column> columns)
    {
        this.columns = columns;
    }

    @JsonValue
    public List<Column> getColumns()
    {
        return columns;
    }

    public int getColumnCount()
    {
        return columns.size();
    }

    public Column getColumn(int index)
    {
        return columns.get(index);
    }

    public String getColumnName(int index)
    {
        return getColumn(index).getName();
    }

    public Type getColumnType(int index)
    {
        return getColumn(index).getType();
    }

    public boolean isEmpty()
    {
        return columns.isEmpty();
    }

    public int size()
    {
        return columns.size();
    }

    public int getFixedStorageSize()
    {
        int total = 0;
        for (Column column : columns) {
            total += column.getType().getFixedStorageSize();
        }
        return total;
    }

    public void consume(RecordCursor cursor, RecordConsumer consumer)
    {
        for(Column c : columns) {
            c.consume(cursor, consumer);
        }
    }

    public void produce(PageBuilder builder, RecordProducer producer)
    {
        for(Column c : columns) {
            c.produce(builder, producer);
        }
    }

    @Override
    public String toString()
    {
        StringBuilder sbuf = new StringBuilder();
        sbuf.append(String.format(" Schema: %d columns", columns.size())).append('\n');
        for (Column c : columns) {
            sbuf.append("   " + c.toString()).append('\n');
        }
        return sbuf.toString();
    }
}

File: lib/quickload/bridge/schema.rb
module QuickLoad
  module Bridge
    class Schema < Struct.new(:columns)
      def initialize(columns)
        super(columns)
      end

      def self.wrap(schema)
        new(schema.getColumns.to_a)
      end

      def to_java
        java_columns = columns.map {|c| c.to_java }
        Java::Schema.new(java_columns)
      end
    end

    class Column < Struct.new(:index, :name, :type)
      def initialize(index, name, type)
        super(index, name, type.getName.to_sym)
        @java_type = type
      end

      attr_reader :java_type

      def to_java
        Java::Column.new(index, name, @java_type)
      end
    end
  end
end
